api/indexer/indexer.go

```diff
@@ -8,230 +8,262 @@ import (
 	"encoding/binary"
 	"errors"
 	"fmt"
+	"math"
 	"path/filepath"
-	"sync/atomic"
+	"sync"
 
 	"github.com/ava-labs/avalanchego/database"
 	"github.com/ava-labs/avalanchego/ids"
-	"github.com/ava-labs/avalanchego/utils/wrappers"
+	"github.com/ava-labs/avalanchego/utils/buffer"
 	"github.com/prometheus/client_golang/prometheus"
 
 	"github.com/ava-labs/hypersdk/chain"
 	"github.com/ava-labs/hypersdk/event"
-	"github.com/ava-labs/hypersdk/fees"
-	"github.com/ava-labs/hypersdk/internal/cache"
 	"github.com/ava-labs/hypersdk/internal/pebble"
 )
 
 const maxBlockWindow uint64 = 1_000_000
 
-var errBlockNotFound = errors.New("block not found")
+var (
+	blockEntryKeyPrefix  = []byte("block")
+	latestBlockHeightKey = []byte("latestblockheight")
+)
+
+var (
+	errBlockNotFound          = errors.New("block not found")
+	errTxResultNotFound       = errors.New("transaction result not found")
+	errZeroBlockWindow        = errors.New("indexer configuration of non zero block window is expected")
+	errInvalidBlockWindowSize = errors.New("invalid indexer block window size specified")
+)
 
 var _ event.Subscription[*chain.ExecutedBlock] = (*Indexer)(nil)
 
 type Indexer struct {
 	blockDB            *pebble.Database // height -> block bytes
-	blockIDToHeight    *cache.FIFO[ids.ID, uint64]
-	blockHeightToBlock *cache.FIFO[uint64, *chain.ExecutedBlock]
-	blockWindow        uint64 // Maximum window of blocks to retain
-	lastHeight         atomic.Uint64
+	cachedBlocks       buffer.Queue[uint64]
+	blockIDToHeight    map[ids.ID]uint64
+	blockHeightToBlock map[uint64]*chain.ExecutedBlock
+	txCache            map[ids.ID]cachedTransaction
+	evictedBlocks      []uint64 // list of blocks that are queued for deletion.
```

Comment:
Since using the bounded buffer forces us to add to this slice via a callback function (bit harder to follow than ideal), could we switch to using a data structure that allows us to easily return the element that should be removed like a queue?